### DISCUSSION TITLE
How to carry URL query parameters from page to page?

### DISCUSSION THREAD

**Discussion by eliorivero** [vercel/next.js#20917] [Help]
Hi, I've an app where my users land on a page that starts a flow. The URL that brings them to this start page has an `id` and a `redirect` query parameters. Once in the page, users can click something and move to the next step and so on.

I'd like to keep the URL query parameters in the URL at all times while the user advances through the pages, without having to add these to each URL in each <Link> instance.

I could use local or session storage, and once users arrive to the page, strip the query params from the URL and save them to local storage and use them from there in the next pages. However, it would be nice if this could be kept within Next.js.

I tried looking into the router events but they're not useful for this case. It would be nice if there was a way to "filter" the URL before the route changes, to append the query params at that moment before proceeding to the next page.

**Comment by leerob** [comment]
Just a thought: have you considered using paths instead of query params? 

`import/step-one/sub-option/another-one` vs. `import?step=one&option=sub&another=one`.

**Reply by eliorivero** [reply]
Yea, I am using paths for another identifier. The issue here is that some of the things I want to pass as query params might not exist every time for these paths and that’s why I haven’t modeled them after paths.
